src/lando/middleware.py

```diff
@@ -105,3 +109,50 @@ def __call__(self, request: WSGIRequest) -> HttpResponse:
             )
 
         return self.get_response(request)
+
+
+class cProfileMiddleware:
+    """A middleware to profile requests/responses and return the result.
+
+    To generate a profile report for a request, the following conditions must be met:
+    - Request must be sent by an authenticated staff user.
+    - The PROFILING_ENABLED configuration variable must be set to True.
+    - The "profile" query parameter must be passed in the URL.
+
+    In addition, the following parameters can be passed to customize the report:
+    - sort (can be set to a column in the report)
+    """
+
+    @staticmethod
+    def _should_profile(request: WSGIRequest) -> bool:
+        return (
+            ConfigurationVariable.get(ConfigurationKey.PROFILING_ENABLED, False)
+            and request.user.is_staff
+            and "profile" in request.GET
+        )
+
+    def __init__(self, get_response: Callable[[HttpRequest], HttpResponse]):
+        self.get_response = get_response
+
+    def __call__(self, request: HttpRequest) -> HttpResponse:
+        return self.get_response(request)
+
+    def process_view(
+        self,
+        request: WSGIRequest,
+        view_func: Callable,
+        view_args: tuple,
+        view_kwargs: dict,
+    ) -> Optional[HttpResponse]:
+        """If profile is requested, run cprofile and generate the output in html format."""
+        if not self._should_profile(request):
+            return
+
+        profiler = cProfile.Profile()
+        profiler.runcall(view_func, request, *view_args, **view_kwargs)
+        profiler.create_stats()
+        out = StringIO()
+        stats = pstats.Stats(profiler, stream=out)
+        stats.strip_dirs().sort_stats(request.GET.get("sort", "cumtime"))
+        stats.print_stats()
+        return HttpResponse(f"<pre>{out.getvalue()}</pre>")
```

Comment:
How confident are we that no sensitive data (e.g., SSH credentials) would get returned in the stats?

Reply:
It just includes a breakdown of function calls (method name is probably the most detail here), without any info on variables or anything else like that.

E.g.,

```
         165062 function calls (163364 primitive calls) in 7.233 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    7.233    7.233 base.py:97(view)
        1    0.000    0.000    7.233    7.233 base.py:134(dispatch)
        1    0.000    0.000    7.233    7.233 revisions.py:69(get)
        3    0.000    0.000    7.231    2.410 auth.py:140(wrapped)
...
```